### PR TITLE
FEAT-#2013: Improved merge_asof implementation

### LINKED ISSUE #2013: Implement distributed merge_asof
Currently `modin.pandas.merge_asof` falls back to Pandas. It would be good to implement a distributed, Modin-native implementation.

First step will be figuring out if this should be implemented in one go, or if it's big enough to merit multiple sub-tasks.

### PR DESCRIPTION
**Pull request by itamarst**
## What do these changes do?

This is a sketch of an approach to `merge_asof` that is more efficient than the current approach, hopefully.

@devin-petersohn would appreciate some feedback on e.g. which bits are potentially inefficient.

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [ ] passes `flake8 modin`
- [ ] passes `black --check modin`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #2013
- [ ] tests added and passing

### CODE REVIEW

**Comment by itamarst** [comment]
@devin-petersohn ping—if you have a chance, would love some feedback on this sketch before proceeding.

**Review by devin-petersohn — COMMENTED** [review]
Sorry for the delay in response, I took some vacation to rest. I'm still getting caught up.

The implementation here, while not super efficient, will be better than the current default behavior. We should still probably throw the warning for defaulting to pandas because we are collecting whole column(s) to the driver to compute the merge.

Otherwise it looks good until we can have a better internal way of joins with conditionals.

**Comment by itamarst** [comment]
Will re-open when it's ready for review.

**Comment by itamarst** [comment]
@devin-petersohn probably still needs testing of some of the fallback edge cases, but this is a lot closer to done, and so could use another passthrough.

So many edge cases... and I feel like I'm handling them in super-hacky ways.

**Comment by itamarst** [comment]
@devin-petersohn any thoughts?

**Review by devin-petersohn — COMMENTED** [review]
This looks like a good approach given the current internals. Eventually, I would like to implement a non-equi join internally, which would fit this operator better.

**Comment by itamarst** [comment]
Looks like:

1. Someone fixed the argument handling in `merge_asof` 17 days ago, so it'll be in Pandas 1.2.0.
2. There's a PR for one of the weird seeming argument combos: https://github.com/pandas-dev/pandas/issues/33463

**Comment by itamarst** [comment]
@devin-petersohn I guess this is ready for review now? Main caveat I have is that example dataframes for tests were taken from Pandas docs, so potentially licensing issue that would need to fixed, not sure whether that's a problem or not.

File: modin/pandas/general.py [hunk 1]
```diff
@@ -158,28 +158,147 @@ def merge_asof(
...
+        left_index=True,
+        right_index=True,
+        suffixes=suffixes,
+        how="left",
     )
 
+    # 5. Clean up to match Pandas output:
+    if left_on is not None and right_index:
+        result.insert(
+            list(result.columns).index(left_on + suffixes[0]),
```
Comment (devin-petersohn) [inline modin/pandas/general.py:293]: This avoids copying into a list
```suggestion
            result.columns.get_indexer_for([left_on + suffixes[0])[0],
```

File: modin/pandas/general.py [hunk 2]
```diff
@@ -158,28 +158,147 @@ def merge_asof(
...
+    # If we haven't set these by now, there's a bug in this function.
+    assert left_column is not None
+    assert right_column is not None
+
+    if by is not None:
+        if left_by is not None or right_by is not None:
+            raise ValueError("Can't have both 'by' and 'left_by' or 'right_by'")
+        left_by = right_by = by
+
+    # List of columsn case should have been handled by direct Pandas fallback
```
Comment (devin-petersohn) [inline modin/pandas/general.py:239]:
```suggestion
    # List of columns case should have been handled by direct Pandas fallback
```

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Looks good, a couple of minor points. Would you be able to fix the failing commitlint test?

**Comment by itamarst** [comment]
I can fix the commit by squash merging, my attempts at rebases have failed due to limits of my Git knowledge. So I can do that, or if you know you can do that. Perhaps the 70 character limit could be removed? It doesn't seem lack it adds much... :grin:

**Comment by devin-petersohn** [comment]
Part of the 70 limit is a github thing, messages longer than that are chopped. I increased it to 88 in #2489.

**Comment by itamarst** [comment]
OK, merged forward, hopefully everything will pass (once GitHub Actions gets unbroken for the day, I guess? Seems like they're having issues).

**Comment by itamarst** [comment]
@devin-petersohn another frustrating git commit failure. I'm going to just turn this into a patch, and create a completely new PR with a new branch :(
